msal_extensions/token_cache.py

```diff
@@ -0,0 +1,196 @@
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            super(FileTokenCache, self).add(event, **kwargs)  # pylint: disable=duplicate-code
+            self._write()
+
+    def modify(self, credential_type, old_entry, new_key_value_pairs=None):
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            super(FileTokenCache, self).modify(
+                credential_type,
+                old_entry,
+                new_key_value_pairs=new_key_value_pairs)
+            self._write()
+
+    def find(self, credential_type, **kwargs):  # pylint: disable=arguments-differ
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            return super(FileTokenCache, self).find(credential_type, **kwargs)
+
+    def __getattr__(self, item):
+        # Instead of relying on implementers remembering to update _last_sync, just detect that one
+        # of the relevant methods has been called take care of it for derived types.
+        if item in ['_read', '_write']:
+            self._last_sync = int(time.time())
+        return super(FileTokenCache, self).__getattr__(item)  # pylint: disable=no-member
+
+
+class WindowsTokenCache(FileTokenCache):
+    """A SerializableTokenCache implementation which uses Win32 encryption APIs to protect your
+    tokens.
+    """
+    def __init__(self,
+                 cache_location=None,
+                 lock_location=None,
+                 entropy=''):
+        super(WindowsTokenCache, self).__init__(
+            cache_location=cache_location,
+            lock_location=lock_location)
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+
+    def _write(self):
+        with open(self._cache_location, 'wb') as handle:
+            handle.write(self._dp_agent.protect(self.serialize()))
+
+    def _read(self):
+        with open(self._cache_location, 'rb') as handle:
+            cipher_text = handle.read()
+        contents = self._dp_agent.unprotect(cipher_text)
+        self.deserialize(contents)
+
+
+class OSXTokenCache(FileTokenCache):
+    """A SerializableTokenCache implementation which uses native Keychain libraries to protect your
+    tokens.
+    """
+
+    def __init__(self,
+                 cache_location='~/.IdentityService/msal.cache',
+                 lock_location=None,
+                 service_name='Microsoft.Developer.IdentityService',
+                 account_name='MSALCache'):
```

Comment:
This one is slightly different than the `WindowsTokenCache.__init__(...)`. The `cache_location` parameter does need to be explicitedly (re)defined here because it would have a different default value. But the `lock_location` can probably still go with the `**kwargs` pattern.

Reply:
I realized that the default value is actually the same as the default value in FileTokenCache. For that reason, the cache_location value specified here is unnecessary.

Fixed in 0327141ce3e32968c30e06ef722e25849782409a